### ISSUE TITLE
[LSP Diagnostics Feature Enhancement] delay_until_save option for on_publish_diagnostics (if possible)

### ISSUE THREAD

**Issue by bluz71** [neovim/neovim#13324]
Hello Neovim LSP folks,

Spawning a new issue from [this Reddit post](https://www.reddit.com/r/neovim/comments/jw9i8b/lsp_diagnostics_is_it_possible_to_only_display).

As an ALE user one can specify to emit diagnostics after certain events such as: on buffer enter, on insert leave, on file save etc.

I use *on file save* since I don't like to see diagnostics until I actually change a file. This makes browsing files, often not my own, far more pleasant since linting/diagnostics display can be distracting. I prefer to see diagnostics only when I am actually editing files.

With linting it is often the case that a separate command line utility is run which makes setting up *on file save* processing much easier (I assume). LSP diagnostics however are produced by the language server even when not asked for, usually after the *documentOpen* request been processed (if I recall correctly).

So I can understand why Neovim LSP diagnostics appear out of thin air soon after opening a buffer.

I do wonder whether it is possible to delay the display of diagnostics until after the first file save? I would find that ideal.

I wonder, and hope, that something like this is possible:

```viml
vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    signs = true,
    underline = false,
    update_in_insert = false,
    virtual_text = true,
    delay_until_save = true <<< NEW OPTION, MAYBE
  }
)
```

This theoretical example states to only display diagnostics after a file save operation has occurred, and then proceed as usual.

Best regards.

**Comment by rockerBOO** [comment]
@tjdevries Having some way to display the diagnostics on demand (signs, virtual_text and popups). Probably need a way to clear diagnostics too if there isn't. Toggling them on/off would be cool too.

**Comment by Iron-E** [comment]
@rockerBOO you can clear diagnostics with:

```lua
vim.lsp.util.buf_clear_diagnostics(0, vim.lsp.util.get_line_diagnostics())
```

But they'll come back on the next save or text change.

**Comment by mjlbach** [comment]
@tjdevries What do you think of letting people just pass the list of autocmd's they want diagnostics to trigger on? Or do you want me to just add this option? I have a local branch that could be cleaned up that does the former.

**Comment by tjdevries** [comment]
Yeah, we can probably just change it to list of autocmds to do the display. It might be slightly tricky to denounce / not call display multiple times, but it should be OK.

**Comment by tjdevries** [comment]
Do you want me to do it?

**Comment by mjlbach** [comment]
Totally your call, I can probably get it done tomorrow.

**Comment by tjdevries** [comment]
OK, we should remove `update_in_insert` as an option and change it to something like `display_autocmds`, and if anything is passed, then we will wait until one of those autocmds happens to display the updates, otherwise (if nothing is passed) then we will always update right away.

That's the better solution and won't explode the complexity. Users can configure the autocmds they want to have it work for.

**Comment by tjdevries** [comment]
(and you can do it and just tag me and I can take a look)

**Comment by tjdevries** [comment]
(oh and normally, we'd leave and suport `update_in_insert` but we haven't released it yet and this is why it's good not to release too early 😆 )

**Comment by mjlbach** [comment]
@bluz71 @rockerBOO @Iron-E Can you all try #13742? There's instructions how to use it in the PR.

**Comment by bluz71** [comment]
@mjlbach,

I am not getting consistent sign / virtual text updating.

My configuration:

```lua
local diagnostic_handler = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    signs = {
      severity_limit = 'Warning',
    },
    underline = false,
    show_diagnostic_autocmds = { 'InsertLeave', 'TextChanged' },
    virtual_text = {
      spacing = 2,
      severity_limit = 'Warning',
    },
  }
)
```

I think `InsertLeave` is being triggered and updated correctly.

But, `TextChanged` is giving me grief. If I delete a line via `dd` (to trigger a diagnostic) I don't get updated signs & virtual text, likewise when I do `u` to undo (and fix). It feels like I am not getting triggering happening.

Cheers.

**Comment by mjlbach** [comment]
Hmm try adding TextYankPost and see if that fixes the former.

**Comment by mjlbach** [comment]
If you delete the text I believe the highlight will go away, but if the autocmd isn't triggered it won't update the signs which might explain the decoupling

**Comment by bluz71** [comment]
> Hmm try adding TextYankPost and see if that fixes the former.

No change.

> If you delete the text I believe the highlight will go away, but if the autocmd isn't triggered it won't update the signs which might explain the decoupling

I am doing the reverse.

I have a error free file at first.
I then do a `dd` to break syntax.
I expect signs and virtual text to be displayed but they do not.

Interestingly if I do `lua vim.lsp.diagnostic.goto_next()` then the diagnostic will be displayed in a floating window. So the diagnostic does exist under the covers but is not being updated after the triggering.

It is my understanding that `dd` should trigger `TextChanged`.

Cheers.

**Comment by mjlbach** [comment]
Yeah, but you have to remember that it will *immediately* trigger display. The language server might not have returned the updated diagnostic by then (and likely hasn't). That's why I suggest the hover  "CursorHoldI"

**Comment by bluz71** [comment]
> Yeah, but you have to remember that it will immediately trigger display. The language server might not have returned the updated diagnostic by then (and likely hasn't). That's why I suggest the hover "CursorHoldI"

So there is a race condition.

When I hit `dd` it triggers a request for diagnostic and immediately triggers `TextChanged`-based `show_diagnostic_autocmds`, which is blank at that instant, but potentially a couple hundred milliseconds later it will be populated (probably asynchronously).

Hmmm, this is problematic for my use case.

I definitely do not want to use `CursorHold` since I really only want to trigger diagnostics when saving a file (which happens automatically for me via the [vim-auto-save](https://github.com/907th/vim-auto-save) triggered by `InsertLeave` and `TextChanged`.

Is there anyway to make the pipeline more synchronous?

**Comment by bluz71** [comment]
> you'll also likely want to lower vim.o.updatetime (I use 250)

I have `set updatetime=100` to make GitGutter more responsive.

**Comment by mjlbach** [comment]
What exact behavior do you want? With the patch in the thread, if you don't pass any autocommands, it will always render diagnostics as soon as they are sent from the server. The majority of the time, the return will be within a few hundred milliseconds. If we were to block until response from the server, that would introduce horrible lag. It's already asychronous, this is just exposing control over when you render the updates.

**Comment by bluz71** [comment]
> What exact behavior do you want? 

In my particular case, I want sign and virtual text display and updating to happen after a response from server has been received. That is if my setup specifically triggers diagnostic requests after certain behaviours.

For any autocmd that triggers an actual diagnostic LSP request then I want a wait to occur before updating signs and virtual text. Or alternatively have sometime of internal callback that notices that the server has sent new diagnostics, triggered by an autocmd of sorts, hence update the display with those new details.

**Comment by mjlbach** [comment]
I think there's a slight misunderstanding. You're always receiving diagnostics from the server, we don't request them. The only thing we do with the autocommands is determine when to show them to you. If you always want diagnostics to display immediately as the handler is invoked, then just use the patch in the PR issue and don't set the autocommands to be anything. I'll probably just set this as the default.

**Comment by mjlbach** [comment]
Here, try not setting any autocommands and pull the latest commit of the PR, I just set the default to what I think you want.

**Comment by bluz71** [comment]
> Here, try not setting any autocommands and pull the latest commit of the PR, I just set the default to what I think you want.

I want diagnostics delayed until at least the first file save and from then on to be updated in real time but explicitly **not** whilst in insert mode (aka I liked the old `￼update_in_insert = false` option).

What do you recommend as the best `show_diagnostic_autocmds` to achieve that for me?

**Comment by bluz71** [comment]
> I want diagnostics delayed until at least the first file save 

You may ask why I want that? As I note in the first post, I often browse other people's code (JavaScript) which 99% of the time does not confirm to my particular `ESLint` rules. The popping of signs and virtual text whilst I am in pure browse code mode is highly annoying and pretty much useless (since I won't be fixing these lint errors).

In [ALE](https://github.com/dense-analysis/ale) one can set `let g:ale_lint_on_save = 1` to prevent diagnostics until actual editing happens.

Just giving background about my particular use case. I suspect a percentage of ALE users who will come over may want to achieve something similar.

**Comment by mjlbach** [comment]
```I want diagnostics delayed until at least the first file save and from then on to be```

We're probably never going to be able to give you that level of granularity. With this method, you can lint after every save with ` BufWritePost`. You'll have to write your own handler if you want bufwrite post to toggle a different set of autocommands.  

> from then on to be updated in real time but explicitly not whilst in insert mode (aka I liked the old ￼update_in_insert = false option).
`InsertLeave`, `TextChanged`, `CursorHold` should do it

**Comment by mjlbach** [comment]
FWIW this PR shouldn't have changed anything, if you use "InsertLeave", "CursorHoldI" the behavior should be exactly as it was prior to this PR.

**Comment by bluz71** [comment]
> We're probably never going to be able to give you that level of granularity. With this method, you can lint after every save with BufWritePost.

Except `BufWritePost` by itself has the race condition discussed above (server response will be delivered after the sign/virt-text display triggering has already occured) which makes it a no go. 

Using `CursorHold` will on the other hand will emit diagnostics automatically after file open (which has been the Neovim LSP behaviour since the start anyway). I actually do want CursorHold triggered, but only after the first file save has occurred. Something that probably won't be implemented.

It is what it is. I will probably stick with [ALE](https://github.com/dense-analysis/ale) via their `lint_on_save` option.

Cheers.

**Comment by mjlbach** [comment]
There isn't a race condition, the difference is that ALE (I believe, reading through their source briefly) not using this option for language servers, but rather for running synchronous linters. The race condition is that there will be a 100-200 ms delay until cursor hold updates your diagnostics, whereas with ALE, the delay will be longer if the linting is only triggered after save.

**Comment by bluz71** [comment]
> The principal difficulty is we don't know when the language server will return all of it's diagnostics. We might be able to optionally add timer start, so that when you trigger the save autocmd, it's delayed 200 ms until the language server is likely up to date.

A user configurable delay would actually solve my issue. I would probably set it to 500-750ms since I don't mind the diagnostics coming in a tad after file save.

**Comment by bluz71** [comment]
> There isn't a race condition, the difference is that ALE (I believe, reading through their source briefly) is not using this option for language servers, but rather for running dispatched linters triggered on save.

I just tested ALE with the Dart Language Server (analysis server) and it definitely does pause diagnostic display until after file save if `lint_on_save` is set. From then on it updates diagnostics only after each save, at least that is what I see.

But Neovim LSP does not need, nor want, to re-implement ALE. If somebody wants ALE they should use ALE.

However, the idea floated of a `delay` would be excellent since it would work well with certain types of autocmds (specifically BufWritePost).

**Comment by mjlbach** [comment]
I'm just going to go ahead  implement the wrapper around the command called via autocmd using timer_start, and I think you'll find the experience is the same as ALE.

**Comment by bluz71** [comment]
Thanks @mjlbach I appreciate that very much.

I would be more than willing to test it out.
I too think a `time_start` with some type of delay would solve my particular use case.

For me I think `show_diagnostic_autocmds = { 'InsertLeave', 'TextChanged' }` with between 300-600ms display delay would be just about ideal.

Good chat today.

**Comment by mjlbach** [comment]
@bluz71 You can try this branch, I don't think it will make it into the PR. For the settings you want, I would start with this:

```
      show_diagnostic_autocmds = { "BufWritePost" },
      diagnostic_delay = 1000
```

**Comment by bluz71** [comment]
It will take me about a day to report back due to nighttime and existing commitments tomorrow morning. But after that it will definitely give it a go.

Stay tuned.

**Comment by bluz71** [comment]
PR #13748 solves this issue.

**Comment by bluz71** [comment]
@mjlbach,

It may be worthwhile, time permitting, to revisit this now that byte_ranges / increment change is temporarily paused.

@tjdevries was not a fan of the API in #13748 as noted by his post.

> Some thoughts (will probably edit some more):
>
> 1. This approach does not automatically share the configuration of the display with the handler
> 2. This approach may lead to calling showing displays much more often than necessary (if you're calling > it in a timer for example, which will be set even when there are no new diagnostics to display).
>
> It is possible to debounce in the old method as well, but perhaps slightly complicated atm. I think we can make it possible to debounce more easily when we ship more debounce utilties / wrappers.

He may prefer your proposal a few posts up, something like:

```lua
vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    signs = true,
    underline = false,    
    virtual_text = true,
    show_diagnostic_autocmds = { 'InsertLeave', 'TextChanged' }, -- NEW OPTION
    diagnostic_delay = 500  -- NEW OPTION
  }
)
```

Most folks wouldn't need to set `show_diagnostic_autocmds` and `diagnostic_delay`, but some who want tight control and when and how to show diagnostic will.

Cheers.

**Comment by mjlbach** [comment]
I actually fixed 2 already, display only gets called on change. 1 I'm not sure, I've discussed the need for a global lsp namespace for settings but not sure what the consensus is

**Comment by tjdevries** [comment]
We should decide on this before 0.5 since it's a large API area.

I can try and look again at the PR and see what I think. I want it to be 1-line for simple situations for users to configure :) so it may be difficult

**Comment by mjlbach** [comment]
I think the way I did it is a decent approach, and with the addition of a namespace for options could be standardized, but you have way more experience with this than me so I'm going to defer to you on this one

**Comment by bluz71** [comment]
> We should decide on this before 0.5 since it's a large API area.

Strong agreement on this.

> I think the challenging part is it seems like people have a lot of different opinions + factorial combinations of when they want diagnostics refreshed.

I think this API takes care of most such combinations:

```lua
vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    signs = true,
    underline = false,    
    virtual_text = true,
    show_diagnostic_autocmds = {'InsertLeave', 'TextChanged'},
    diagnostic_delay = 500
  }
)
```

User get to chooses after which Neovim events diagnostics will be updated on screen and they get to factor in a small delay to give the language server time to send over the diagnostics (for example after a text edit). 

Heck I would really like `diagnostic_delay` not to exist and for the Neovim LSP client to factor in a delay all by itself if a user has set `show_diagnostic_autocmds`. What value to choose? I just looked it up, [ALE](https://github.com/dense-analysis/ale/blob/master/plugin/ale.vim#L71) appears to use a 200ms delay, but I am not sure that applies to language servers as compared to command line linters. From my experiments 200ms was not enough time. A user override probably will be needed since this is a *how long is a piece of string issue*.

**Comment by mjlbach** [comment]
That was my first pass implementation ^, it wasn't flexible enough to allow things like delays, which is why I changed it

**Comment by bluz71** [comment]
Sorry for the bump, just wanting to get a feel for where we are and the path ahead.

PR #13748 solves this issue, but @tjdevries was not entirely happy with the API as [noted here](https://github.com/neovim/neovim/pull/13748#issuecomment-760288071). I can understand where he is coming from. I am not too fussed with whatever API is chosen just as long as it achieves the desire of this issue, that being allowing users to convert asynchronous diagnostics (LSP default) to synchronous event-driven diagnostics (aka like ALE can do).

So the current state of this issue is paused due to other more pressing matters, including the 0.5 release which does feel somewhat close.

If this does not make the cut for 0.5, and it probably won't, then may I ask that it be considered for 0.5.1 or 0.5.2 rather than 0.6? For example `debounce_did_change_notify` #14119 looks like it will be 0.5.1.

Cheers.

**Comment by bluz71** [comment]
I no longer need this.

I have setup global diagnostics toggling as noted in [this Reddit post](https://www.reddit.com/r/neovim/comments/ng0dj0/lsp_diagnostics_query_is_there_an_way_to_toggle).

I have bound that to a mapping which I toggle when I now longer want to see signs/virtual-text diagnostics as noted by [this post](https://www.reddit.com/r/neovim/comments/ng0dj0/lsp_diagnostics_query_is_there_an_way_to_toggle/gyoogr2/). That achieves what I desire.

From my perspective this issue can be closed.

**Comment by lcrockett** [comment]
If it helps, just pitching in for anyone ending up here after trying to figure out (via GitHub, Reddit, Google, etc.) how to selectively and permanently enable / disable (parts) of LSP indicators, such as signs and virtual text or trying to accomplish similar end results using `autocmd` (for instance, to display indicators only after saving); see below for a few 'creative' snippets that accomplishes this goal.

The result of using buffer-local variables in the snippets below is that whichever combination of indicator displaying you choose, it's buffer specific. Changing the buffer-local variables to global variables can easily change this behaviour to allow global changes.

<details>
<summary>file :: ~/.config/nvim/lua/function/function/lsp_diagnostic_indicators.lua :: loaded on-demand via require statement</summary>

```lua
local lsp_diagnostic_indicators = {}

lsp_diagnostic_indicators.none = function()
  _G.lsp_diagnostic_indicators_clear()

  vim.b.lsp_diagnostic_indicators_show = false

  return
end

lsp_diagnostic_indicators.redraw = function()
  _G.lsp_diagnostic_indicators_redraw()

  return
end

lsp_diagnostic_indicators.signs = function()
  _G.lsp_diagnostic_indicators_signs()
  _G.lsp_diagnostic_indicators_redraw()

  vim.b.lsp_diagnostic_indicators_mode = 'Signs'

  if not vim.b.lsp_diagnostic_indicators_show then
    vim.b.lsp_diagnostic_indicators_show = true
  end

  return
end

lsp_diagnostic_indicators.signsvirtualtext = function()
  _G.lsp_diagnostic_indicators_signsvirtualtext()
  _G.lsp_diagnostic_indicators_redraw()

  vim.b.lsp_diagnostic_indicators_mode = 'SignsVirtualText'
  vim.b.lsp_diagnostic_indicators_show = true

  if not vim.b.lsp_diagnostic_indicators_show then
    vim.b.lsp_diagnostic_indicators_show = true
  end

  return
end

lsp_diagnostic_indicators.virtualtext = function()
  _G.lsp_diagnostic_indicators_virtualtext()
  _G.lsp_diagnostic_indicators_redraw()

  vim.b.lsp_diagnostic_indicators_mode = 'VirtualText'
  vim.b.lsp_diagnostic_indicators_show = true

  if not vim.b.lsp_diagnostic_indicators_show then
    vim.b.lsp_diagnostic_indicators_show = true
  end

  return
end

return lsp_diagnostic_indicators
```
</details>

<details>
<summary>file :: ~/.config/nvim/lua/plugin/nvim-lspconfig.lua :: loaded via tjdevries/astronauta.nvim</summary>

```lua
local on_attach = function(bufnr)
  local function buf_set_option(...) vim.api.nvim_buf_set_option(bufnr, ...) end

  buf_set_option('omnifunc', 'v:lua.vim.lsp.omnifunc')
end

require('lspconfig').lua.setup {
  autostart = true,
  settings = {
    Lua = {
      diagnostics = {
        enable = true,
        globals = {
          'P',
          'after_each',
          'awesome',
          'before_each',
          'client',
          'describe',
          'it',
          'theme',
          'vim'
        }
      },
      telemetry = {
        enable = false
      }
    }
  },
  on_attach = on_attach
}

vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    severity_sort = true,
    signs = false,
    underline = false,
    update_in_insert = false,
    virtual_text = false
  }
)

vim.fn.sign_define {
  { name = 'LspDiagnosticsSignError',       text = '', texthl = 'LspDiagnosticsSignError' },
  { name = 'LspDiagnosticsSignHint',        text = '', texthl = 'LspDiagnosticsSignHint' },
  { name = 'LspDiagnosticsSignInformation', text = '', texthl = 'LspDiagnosticsSignInformation' },
  { name = 'LspDiagnosticsSignWarning',     text = '', texthl = 'LspDiagnosticsSignWarning' }
}

vim.lsp.set_log_level(vim.lsp.log_levels.WARN)

function _G.lsp_diagnostic_indicators_signs()
  vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
    vim.lsp.diagnostic.on_publish_diagnostics, {
      severity_sort = true,
      signs = true,
      underline = false,
      update_in_insert = false,
      virtual_text = false
    }
  )
end

function _G.lsp_diagnostic_indicators_signsvirtualtext()
  vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
    vim.lsp.diagnostic.on_publish_diagnostics, {
      severity_sort = true,
      signs = true,
      underline = false,
      update_in_insert = false,
      virtual_text = {
        severity_limit = 'Hint',
        spacing = 10
      }
    }
  )
end

function _G.lsp_diagnostic_indicators_virtualtext()
  vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
    vim.lsp.diagnostic.on_publish_diagnostics, {
      severity_sort = true,
      signs = false,
      underline = false,
      update_in_insert = false,
      virtual_text = {
        severity_limit = 'Hint',
        spacing = 10
      }
    }
  )
end

function _G.lsp_diagnostic_indicators_redraw()
  -- NOTE: This creative monstrous function might become obsolete after the merge of 'https://github.com/neovim/neovim/pull/13748', who knows !
  vim.lsp.handlers["textDocument/publishDiagnostics"](
    nil,
    "textDocument/publishDiagnostics", {
        diagnostics = vim.lsp.diagnostic.get(0, tonumber(vim.inspect(vim.tbl_keys(vim.lsp.buf_get_clients())):match("%{ (.-) %}"))),
        uri = vim.uri_from_bufnr(0)
    },
    tonumber(vim.inspect(vim.tbl_keys(vim.lsp.buf_get_clients())):match("%{ (.-) %}"))
  )
end

function _G.lsp_diagnostic_indicators_clear()
  vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
    vim.lsp.diagnostic.on_publish_diagnostics, {
      severity_sort = true,
      signs = false,
      underline = false,
      update_in_insert = false,
      virtual_text = false
    }
  )

  vim.lsp.diagnostic.clear(0)
end
```
</details>

<details>
<summary>usable calls</summary>

```lua
lua require("function.lsp_diagnostic_indicators").none()
lua require("function.lsp_diagnostic_indicators").redraw()
lua require("function.lsp_diagnostic_indicators").signs()
lua require("function.lsp_diagnostic_indicators").signsvirtualtext()
lua require("function.lsp_diagnostic_indicators").virtualtext()
```
</details>

<details>
<summary>file :: ~/.config/nvim/lua/plugin/autocommands.lua :: loaded via tjdevries/astronauta.nvim</summary>

```lua
vim.api.nvim_exec([[
  augroup GlobalBufferVisuals
    autocmd!

    autocmd BufEnter,BufNew,BufHidden * lua require('function.global_buffer_visuals').enable()
    autocmd BufLeave                  * lua require('function.global_buffer_visuals').disable()
  augroup END
]], false)
```
</details>

<details>
<summary>file :: ~/.config/nvim/lua/function/global_buffer_visuals.lua :: called by autocmd group `GlobalBufferVisuals`</summary>

```lua
local global_buffer_visuals = {}

local global_buffer_visuals_exclude_buftype = {
  'nofile',
  'nowrite',
  'prompt',
  'quickfix',
  'terminal'
}

local global_buffer_visuals_exclude_filetype = {
  'Mundo',
  'MundoDiff',
  'NvimTree',
  'TelescopePrompt',
  'fugitive',
  'fugitiveblame',
  'gitmessengerpopup',
  'help',
  'lspinfo',
  'markdown',
  'minimap',
  'packer',
  'qf',
  'tabman',
  'tagbar',
  'toggleterm'
}

global_buffer_visuals.disable = function()
  for _, excluded_buftype in pairs(global_buffer_visuals_exclude_buftype) do
    if vim.bo.buftype == excluded_buftype then
      return
    end
  end

  for _, excluded_filetype in pairs(global_buffer_visuals_exclude_filetype) do
    if vim.bo.filetype == excluded_filetype then
      return
    end
  end

  if vim.b.lsp_diagnostic_indicators_show then
    if #vim.tbl_keys(vim.lsp.buf_get_clients()) >= 1 then
      _G.lsp_diagnostic_indicators_clear()
    end
  end

  return
end

global_buffer_visuals.enable = function()
  for _, excluded_buftype in pairs(global_buffer_visuals_exclude_buftype) do
    if vim.bo.buftype == excluded_buftype then
      return
    end
  end

  for _, excluded_filetype in pairs(global_buffer_visuals_exclude_filetype) do
    if vim.bo.filetype == excluded_filetype then
      return
    end
  end

  if #vim.tbl_keys(vim.lsp.buf_get_clients()) >= 1 then
    if vim.b.lsp_diagnostic_indicators_show == nil then
      vim.b.lsp_diagnostic_indicators_show = true
    end

    if vim.b.lsp_diagnostic_indicators_mode == nil then
      vim.b.lsp_diagnostic_indicators_mode = 'SignsVirtualText'
    end

    if vim.b.lsp_diagnostic_indicators_show then
      if vim.b.lsp_diagnostic_indicators_mode == 'SignsVirtualText' then
        _G.lsp_diagnostic_indicators_signsvirtualtext()
      elseif vim.b.lsp_diagnostic_indicators_mode == 'Signs' then
        _G.lsp_diagnostic_indicators_signs()
      elseif vim.b.lsp_diagnostic_indicators_mode == 'VirtualText' then
        _G.lsp_diagnostic_indicators_virtualtext()
      end

      _G.lsp_diagnostic_indicators_redraw()
    end
  end

  return
end

return global_buffer_visuals
```
</details>

<details>
<summary>example keybinding snippet using folke/which-key.nvim</summary>

```lua
l = {
  name = "+LSP",
  a = { '<cmd>lua require("function.lsp_diagnostic_indicators").signsvirtualtext()<cr>', 'All indicators' },
  n = { '<cmd>lua require("function.lsp_diagnostic_indicators").none()<cr>', 'No indicators' },
  r = { '<cmd>lua require("function.lsp_diagnostic_indicators").redraw()<cr>', 'Redraw indicators' },
  s = { '<cmd>lua require("function.lsp_diagnostic_indicators").signs()<cr>', 'Sign indicators only' },
  v = { '<cmd>lua require("function.lsp_diagnostic_indicators").virtualtext()<cr>', 'Virtual text indicators only' }
}
```
</details>

**Comment by emr-arvig** [comment]
I will just add that having the `show_diagnostic_autocmds` option is still really good for a lot of people and really prevents it from getting super complex like a lot of the other solutions here are.

**Comment by bluz71** [comment]
Associating diagnostics with autocmds fundamentally means converting something that is asynchronous into something that becomes synchronous.

API's to achieve that are complex.
Timing is tricky.
And I am not sure there is much user demand.

My original request spawned from my desire to not be distracted by LSP diagnostics when in browsing mode.

Issue #14825, *add helper to toggle display of diagnostics*, is the solution I am using now and I feel that it is the correct solution for this issue.

Hence, I am closing this since even if something was implemented, I would not use it.

The associated PR may also want to be closed (depending how @mjlbach feels): #13748

**Comment by ddzero2c** [comment]
###

**Comment by jgonera** [comment]
Just wanted to share my solution for those who still would like all diagnostics to remain hidden before a file is saved:

```lua
vim.api.nvim_create_autocmd({"BufNew", "InsertEnter"}, {
-- or vim.api.nvim_create_autocmd({"BufNew", "TextChanged", "TextChangedI", "TextChangedP", "TextChangedT"}, {
  callback = function(args)
    vim.diagnostic.disable(args.buf)
  end
})

vim.api.nvim_create_autocmd({"BufWrite"}, {
  callback = function(args)
    vim.diagnostic.enable(args.buf)
  end
})
```

This disables the diagnostics when a new buffer is created and when you change mode to Insert and re-enables them when you save the file. It doesn't cover some corner cases (e.g. deleting in Normal mode, undo/redo), but works well enough for me 99% of the time without diagnostics jumping around when I'm actually editing code.

You could try replacing `InsertEnter` with a series of [`TextChanged*` events](https://neovim.io/doc/user/autocmd.html#autocmd-events) for a more robust algorithm but it could have performance implications (I'm trying it out now, not sure yet).

**Comment by vovw** [comment]
the solution to this problem is still very hacky, isn't there a inbuilt way to specify for LSP to show diagnostics on file save ??
